File: CPP/merge_sort.cpp
#include <bits/stdc++.h> 
/*Libraries: iostream, fstream, string, vector, sstream
NOTE: PLEASE USE G++ COMPILER OR U MIGHT NOT RUN THIS PROGRAM*/
using namespace std;

struct Record {
    int key;
    string value;
};

vector<string> split(const string& str) {
    vector<string> parts;
    stringstream ss(str);
    string part;
    
    // Get the integer part before the first comma
    getline(ss, part, ',');
    parts.push_back(part); 
    
    // Get the string part (rest of the line)
    getline(ss, part);     
    parts.push_back(part);
    return parts;
}

vector<Record> read_dataset(const string& filename) {
    vector<Record> records;
    ifstream file(filename);
    string line;

    while (getline(file, line)) {
        if (line.empty()) continue;
        stringstream ss(line);
        string key_str, value;
        if (getline(ss, key_str, ',') && getline(ss, value)) {
            try {
                int key = stoi(key_str);
                records.push_back({key, value});
            } catch (...) {
                cerr << "Skipping malformed line: " << line << endl;
            }
        }
    }
    return records;
}

void save_dataset(const vector<Record>& dataset, const string& filename) {
    ofstream file(filename); // open file in write mode
    for (size_t i = 0; i < dataset.size(); ++i) {
        file << dataset[i].key << "," << dataset[i].value;
        if (i + 1 != dataset.size()) file << '\n'; // Add newline for all but the last line
    }
}

void merge(vector<Record>& data, vector<Record>& temp, int left, int mid, int right) {
    //More efficient compared to creating new vector in like lecture slides
	int i = left, j = mid + 1, k = left;
    while (i <= mid && j <= right) {
        if (data[i].key <= data[j].key)
            temp[k++] = data[i++];
        else
            temp[k++] = data[j++];
    }
    while (i <= mid) temp[k++] = data[i++];
    while (j <= right) temp[k++] = data[j++];
    for (int l = left; l <= right; ++l) data[l] = temp[l];
}

// The recursive merge_sort function
void merge_sort(vector<Record>& data, vector<Record>& temp, int left, int right) {
    if (left < right) {//left is basically the start, 0 and right is the end, so in 
                        //most of the case right will always be bigger than
                        //left.. unless they're equal due to the mid equation
        int mid = (left + right) / 2; //so if the total is 5 it becomes 5/2 = 2 (cpp ignores decimals) 
        merge_sort(data, temp, left, mid);
        merge_sort(data, temp, mid + 1, right);
        merge(data, temp, left, mid, right);
    }
}

int main() {
    string dataset_input_filename;
    cout << "Enter dataset filename (e.g., dataset_1000000.csv): ";
    cin >> dataset_input_filename;

    // Read the dataset
    vector<Record> dataset = read_dataset(dataset_input_filename);
    
    if (dataset.empty()) {
        cerr << "No data read. Exiting." << endl;
        return 1;
    }

    // Prepare output filename (e.g., merge_sort_1000000.csv)
    // Find the position of "dataset_" and ".csv" to replace dynamically
    string output_filename_base = dataset_input_filename;
    size_t dataset_pos = output_filename_base.find("dataset_");
    if (dataset_pos != string::npos) {
        output_filename_base.replace(dataset_pos, string("dataset_").length(), "merge_sort_");
    } else {
        // Fallback if "dataset_" is not found, just prepend "merge_sort_"
        output_filename_base = "merge_sort_" + output_filename_base;
    }
    string output_filename = output_filename_base; // This will likely be "merge_sort_N.csv"

    cout << "Dataset size: " << dataset.size() << " elements." << endl;

	vector<Record> temp(dataset.size());

    // Measure running time
    auto start_timer = chrono::high_resolution_clock::now();

    merge_sort(dataset, temp, 0, dataset.size() - 1);

    auto end_timer = chrono::high_resolution_clock::now();
    auto duration = chrono::duration_cast<chrono::milliseconds>(end_timer - start_timer);
    cout << "Merge sort running time: " << duration.count() << " ms" << endl;

    // Save the sorted dataset
    save_dataset(dataset, output_filename);
    cout << "Sorted data saved to " << output_filename << endl;

    return 0;
}

File: py/merge_sort.py
import csv
import time

def read_dataset(filename):
    """Reads a slice of the dataset, returning a list of (int, str) tuples."""
    data = []
    with open(filename, 'r') as f:
        reader = csv.reader(f)
        for row in reader:
            if row:
                data.append((int(row[0]), row[1]))
    return data

def write_sorted_data(filename, sorted_data):
    """Writes the sorted data to a file."""
    with open(filename, 'w', newline='') as f:
        writer = csv.writer(f)
        for item in sorted_data:
            writer.writerow(item)

def merge_sort(dataset, left, right, steps):
    #Basically this keeps seperating the dataset to little segments
    #Afterwards begin merging subarrays from the bottom up
    if left < right:
        mid = (left + right) // 2
        merge_sort(dataset, left, mid, steps)
        merge_sort(dataset, mid + 1, right, steps)
        merge(dataset, left, mid, right, steps)

def merge(dataset, left, mid, right, steps):
    #This arranges the individual element/sorted segments into left and right
    #Then immediately write them back to the dataset
    L_half = dataset[left:mid + 1]
    R_half = dataset[mid + 1:right + 1]

    i = j = 0
    k = left

    while i < len(L_half) and j < len(R_half):
        if L_half[i][0] <= R_half[j][0]:
            dataset[k] = L_half[i]
            i += 1
        else:
            dataset[k] = R_half[j]
            j += 1
        k += 1

    while i < len(L_half):
        dataset[k] = L_half[i]
        i += 1
        k += 1
    while j < len(R_half):
        dataset[k] = R_half[j]
        j += 1
        k += 1

    # Step tracking preserved, though not written out
    steps.append(list(dataset))

if __name__ == "__main__":
    dataset_filename = input("Enter dataset filename (e.g., dataset_1000000.csv): ")
    output_filename_base = dataset_filename.replace("dataset_", "merge_sort_").replace(".csv", "")

    data = read_dataset(dataset_filename)
    steps = [list(data)]  # Initial step

    start_time = time.perf_counter()  # Exclude I/O time
    merge_sort(data, 0, len(data) - 1, steps)
    end_time = time.perf_counter()

    running_time = end_time - start_time
    print(f"Merge sort running time: {running_time:.6f} seconds")

    output_filename = f"{output_filename_base}.csv"
    write_sorted_data(output_filename, data)
    print(f"Sorted data saved to {output_filename}")
